src/System.Memory/src/System/SpanSortHelpers.Keys.cs

```diff
@@ -0,0 +1,234 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+#if !netstandard
+using Internal.Runtime.CompilerServices;
+#endif
+
+using static System.SpanSortHelpersCommon;
+using S = System.SpanSortHelpersKeys;
+using SC = System.SpanSortHelpersKeys_Comparer;
+using SDC = System.SpanSortHelpersKeys_DirectComparer;
+
+namespace System
+{
+    internal static partial class SpanSortHelpersKeys
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Sort<TKey>(this Span<TKey> keys)
+        {
+            int length = keys.Length;
+            if (length < 2)
+                return;
+
+            // PERF: Try specialized here for optimal performance
+            // Code-gen is weird unless used in loop outside
+            if (!SDC.TrySortSpecialized(
+                ref MemoryMarshal.GetReference(keys),
+                length))
+            {
+                DefaultSpanSortHelper<TKey>.s_default.Sort(
+                    ref MemoryMarshal.GetReference(keys),
+                    length);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Sort<TKey, TComparer>(
+            this Span<TKey> keys, TComparer comparer)
+            where TComparer : IComparer<TKey>
+        {
+            int length = keys.Length;
+            if (length < 2)
+                return;
+
+            DefaultSpanSortHelper<TKey, TComparer>.s_default.Sort(
+                ref MemoryMarshal.GetReference(keys),
+                length, comparer);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Sort<TKey>(
+            this Span<TKey> keys, Comparison<TKey> comparison)
+        {
+            int length = keys.Length;
+            if (length < 2)
+                return;
+
+            DefaultSpanSortHelper<TKey>.s_default.Sort(
+                ref MemoryMarshal.GetReference(keys),
+                length, comparison);
+        }
+
+        internal static class DefaultSpanSortHelper<TKey>
+        {
+            internal static readonly ISpanSortHelper<TKey> s_default = CreateSortHelper();
+
+            private static ISpanSortHelper<TKey> CreateSortHelper()
+            {
+                if (typeof(IComparable<TKey>).IsAssignableFrom(typeof(TKey)))
+                {
+                    // coreclr uses RuntimeTypeHandle.Allocate
```

Comment:
Now that lives in coreclr, is it worth having an ifdef?

Reply:
Probably, should be faster, have to see how this can be done in coreclr I guess.